apps/dashr-svm/utils/helperFunctions.R

```diff
@@ -0,0 +1,331 @@
+  if (kernel == "rbf"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "rbfdot",
+      C = C,
+      kpar = list(sigma = sigma),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "linear"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "vanilladot",
+      C = C,
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "poly"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "polydot",
+      C = C,
+      kpar = list(degree = degree, scale = sigma, offset = 0),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "sigmoid"){
+    # tanh kernel equivalent to sklearn's sigmoid kernel (with correct settings)
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "tanh",
+      C = C,
+      kpar = list(scale = sigma, offset = 0),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  }
+  return(svm)
+}
+
+# Plot the datasets and predictions
+generate_main_plot <- function(data, model, threshold){
+  train_data <- data[["train"]]
+  test_data <- data[["test"]]
+  full_data <- rbindlist(data)
+  y_pred_train <- as.factor(
+    as.integer(
+      predict(
+        model,
+        as.matrix(train_data[, c("X1", "X2")]),
+        type = "decision"
+      ) > threshold
+    )
+  )
+  trainConfMatrix <- confusionMatrix(
+    data = y_pred_train,
+    reference = train_data$Y
+  )
+  train_accuracy <- trainConfMatrix$overall["Accuracy"]
+  y_pred_test <- as.factor(
+    as.integer(
+      predict(
+        model,
+        as.matrix(test_data[, c("X1", "X2")]),
+        type = "decision"
+      ) > threshold
+    )
+  )
+  testConfMatrix <- confusionMatrix(
+    data = y_pred_test,
+    reference = test_data$Y
+  )
+  test_accuracy <- testConfMatrix$overall["Accuracy"]
+  rngX1 <- range(full_data$X1) * 1.05
+  rngX2 <- range(full_data$X2) * 1.05
+  grid <- expand.grid(
+    X1 = seq(rngX1[1], rngX1[2], length = 50),
+    X2 = seq(rngX2[1], rngX2[2], length = 50)
+  )
+  grid$decision <- predict(model, grid, type = "decision")[, 1]
+  scaled_threshold <- threshold *
+                     (max(grid$decision) - min(grid$decision)) +
+                     min(grid$decision)
+  r <- max(
+    abs(scaled_threshold - min(grid$decision)),
+    abs(scaled_threshold - max(grid$decision))
+  )
+  csSeq <- seq(0, 1, length.out = 8)
+  colorscale <- list(
+    c(csSeq[1], "#ff744c"),
+    list(csSeq[2], "#ff906d"),
+    list(csSeq[3], "#ffc0a9"),
+    list(csSeq[4], "#ffe7dc"),
+    list(csSeq[5], "#e4fcff"),
+    list(csSeq[6], "#c7feff"),
+    list(csSeq[7], "#97f8ff"),
+    list(csSeq[8], "#00e6ff")
+  )
```

Comment:
I am guessing the `colorscale` should a list nested with lists
`colorscale <- list(
    list(csSeq[1], "#ff744c"),
    list(csSeq[2], "#ff906d"),
    list(csSeq[3], "#ffc0a9"),
    list(csSeq[4], "#ffe7dc"),
    list(csSeq[5], "#e4fcff"),
    list(csSeq[6], "#c7feff"),
    list(csSeq[7], "#97f8ff"),
    list(csSeq[8], "#00e6ff")
  )`

Reply:
That was my first attempt, which does not actually work. For some reason the only way I found it to work was to have the first member of the list as a vector with `c(...)`, and the rest as regular `list`s

Reply:
@sacul-git Weird. I wonder if this comment is informative about the desired structure?

https://github.com/ropensci/plotly/blob/c1d74bffe70f98533ac55ac2bae1dbee461ed966/R/utils.R#L674-L676

Also, seems to mirror the usage here:

https://github.com/ropensci/plotly/blob/02c9aa4915d3f0855fd9b654ef40fa8e9cdaa1be/tests/testthat/test-plotly-colorscale.R#L61-L62

Reply:
Strange, I could have sworn it didn't work when I tried that in the past. Seems to be fine now :man_facepalming: